File: flask_backend/routes/attendance.py
from flask import Blueprint, request, jsonify
from datetime import datetime
from database.db import get_db, write_db

attendance_bp = Blueprint('attendance', __name__)

def next_id(arr):
    return max([e.get('id', 0) for e in arr], default=0) + 1

@attendance_bp.route('/', methods=['GET'])
def get_attendance():
    try:
        db = get_db()
        date = request.args.get('date')
        department = request.args.get('department')
        employee_id = request.args.get('employee_id')
        
        logs = list(db['attendance'])
        
        if date:
            logs = [a for a in logs if a['date'] == date]
        if employee_id:
            logs = [a for a in logs if a['employee_id'] == employee_id]
        if department:
            dept_emp_ids = {e['employee_id'] for e in db['employees'] if e.get('department') == department}
            logs = [a for a in logs if a['employee_id'] in dept_emp_ids]
            
        def sort_key(a):
            # Sort by date desc, then by id desc for consistent ordering
            return (a['date'], a.get('id', 0))
            
        logs.sort(key=sort_key, reverse=True)
        return jsonify({"success": True, "data": logs, "count": len(logs)})
    except Exception as e:
        print(e)
        return jsonify({"success": False, "message": "Failed to fetch attendance"}), 500

@attendance_bp.route('/scan', methods=['POST'])
def mark_attendance():
    try:
        data = request.json
        if not data or not data.get('token'):
            return jsonify({"success": False, "message": "Invalid QR code"}), 400
            
        db = get_db()
        token = data['token']
        import urllib.parse
        clean_token = urllib.parse.unquote(token).strip()
        if clean_token.lower().endswith('.html'):
            clean_token = clean_token[:-5]
            
        def match_emp(e):
            if e.get('secure_token') == token: return True
            if e.get('employee_id', '').lower() == clean_token.lower(): return True
            # Match name ignoring case and spacing
            e_name = e.get('name', '').lower()
            c_name = clean_token.lower()
            if e_name == c_name: return True
            if e_name.replace(' ', '_') == c_name: return True
            if e_name.replace(' ', '') == c_name: return True
            return False
            
        emp = next((e for e in db['employees'] if match_emp(e)), None)
        if not emp:
            return jsonify({"success": False, "message": "Unauthorised or Invalid QR"}), 401
            
        now = datetime.now()
        today = now.strftime('%Y-%m-%d')
        time_in = now.strftime('%I:%M %p')
        
        existing = next((a for a in db['attendance'] if a['employee_id'] == emp['employee_id'] and a['date'] == today), None)
        if existing:
            if 'scans' not in existing:
                existing['scans'] = [existing['time_in']]
            existing['scans'].append(time_in)
            is_check_in = len(existing['scans']) % 2 != 0
            if not is_check_in:
                existing['time_out'] = time_in
            write_db()
            
            return jsonify({
                "success": True,
                "action": 'check_in' if is_check_in else 'check_out',
                "message": f"{'Check-in' if is_check_in else 'Check-out'} recorded for {emp['name']}",
                "data": {
                    "name": emp['name'],
                    "employee_id": emp['employee_id'],
                    "designation": emp.get('designation'),
                    "department": emp.get('department'),
                    "time_in": existing['time_in'],
                    "time_out": existing['time_out'],
                    "date": today
                }
            }), 200

        record = {
            "id": next_id(db['attendance']),
            "employee_id": emp['employee_id'],
            "name": emp['name'],
            "date": today,
            "time_in": time_in,
            "time_out": None,
            "scans": [time_in],
            "marked_by": "QR_SCANNER"
        }
        
        db['attendance'].append(record)
        write_db()
        
        return jsonify({
            "success": True,
            "action": "check_in",
            "message": f"Check-in recorded for {emp['name']}",
            "data": record
        }), 201
    except Exception as e:
        print(e)
        return jsonify({"success": False, "message": "Server error during scan"}), 500

File: backend/routes/attendance.js
const express = require('express');
const router = express.Router();
const { getDB } = require('../database/db');

// ─── POST /api/attendance/scan ────────────────────────────────────────────────
router.post('/scan', (req, res) => {
  try {
    const db = getDB();
    const { token } = req.body;
    if (!token) return res.status(400).json({ success: false, message: 'Token required' });

    // Match by secure_token first, then employee_id, then name
    let employee = db.prepare('SELECT * FROM employees WHERE secure_token = ?').get(token);

    if (!employee) {
      let cleanToken = decodeURIComponent(token).trim();
      if (cleanToken.toLowerCase().endsWith('.html')) cleanToken = cleanToken.slice(0, -5);
      const cName = cleanToken.toLowerCase();

      // Try employee_id match
      employee = db.prepare('SELECT * FROM employees WHERE LOWER(employee_id) = ?').get(cName);

      // Try name match
      if (!employee) {
        const all = db.prepare('SELECT * FROM employees').all();
        employee = all.find(e => {
          const eName = (e.name || '').toLowerCase();
          return eName === cName || eName.replace(/ /g, '_') === cName || eName.replace(/ /g, '') === cName;
        });
      }
    }

    if (!employee) return res.status(404).json({ success: false, message: 'Invalid QR token' });

    const today  = new Date().toISOString().split('T')[0];
    const timeIn = new Date().toLocaleTimeString('en-IN', { hour12: true });

    const existing = db.prepare('SELECT * FROM attendance WHERE employee_id = ? AND date = ?').get(employee.employee_id, today);

    if (existing) {
      const scans = JSON.parse(existing.scans || '[]');
      scans.push(timeIn);
      const isCheckIn = scans.length % 2 !== 0;
      const newTimeOut = isCheckIn ? existing.time_out : timeIn;

      db.prepare('UPDATE attendance SET scans = ?, time_out = ? WHERE employee_id = ? AND date = ?')
        .run(JSON.stringify(scans), newTimeOut, employee.employee_id, today);

      return res.json({
        success: true,
        action: isCheckIn ? 'check_in' : 'check_out',
        message: `${isCheckIn ? 'Check-in' : 'Check-out'} recorded for ${employee.name}`,
        data: {
          name: employee.name,
          employee_id: employee.employee_id,
          designation: employee.designation,
          department: employee.department,
          time_in: existing.time_in,
          time_out: newTimeOut,
          date: today,
        }
      });
    }

    // New attendance record for today
    db.prepare(`
      INSERT INTO attendance (employee_id, date, time_in, time_out, scans, marked_by)
      VALUES (?, ?, ?, NULL, ?, 'QR_SCANNER')
    `).run(employee.employee_id, today, timeIn, JSON.stringify([timeIn]));

    res.json({
      success: true,
      action: 'check_in',
      message: `Check-in recorded for ${employee.name}`,
      data: {
        name: employee.name,
        employee_id: employee.employee_id,
        designation: employee.designation,
        department: employee.department,
        time_in: timeIn,
        time_out: null,
        date: today,
      },
    });
  } catch (err) {
    console.error(err);
    res.status(500).json({ success: false, message: 'Failed to mark attendance' });
  }
});

// ─── GET /api/attendance ──────────────────────────────────────────────────────
router.get('/', (req, res) => {
  try {
    const db = getDB();
    const { date, employee_id } = req.query;

    let sql = `
      SELECT a.*, e.name, e.designation, e.department
      FROM attendance a
      LEFT JOIN employees e ON e.employee_id = a.employee_id
      WHERE 1=1
    `;
    const params = [];

    if (date)        { sql += ' AND a.date = ?';        params.push(date); }
    if (employee_id) { sql += ' AND a.employee_id = ?'; params.push(employee_id); }
    sql += ' ORDER BY a.date DESC, a.id DESC';

    const records = db.prepare(sql).all(...params).map(r => ({
      ...r, scans: JSON.parse(r.scans || '[]')
    }));

    res.json({ success: true, data: records, count: records.length });
  } catch (err) {
    res.status(500).json({ success: false, message: 'Failed to fetch attendance' });
  }
});

// ─── GET /api/attendance/today ────────────────────────────────────────────────
router.get('/today', (req, res) => {
  try {
    const db = getDB();
    const today = new Date().toISOString().split('T')[0];

    const records = db.prepare(`
      SELECT a.*, e.name, e.designation, e.department
      FROM attendance a
      LEFT JOIN employees e ON e.employee_id = a.employee_id
      WHERE a.date = ?
      ORDER BY a.time_in ASC
    `).all(today).map(r => ({ ...r, scans: JSON.parse(r.scans || '[]') }));

    res.json({
      success: true,
      data: records,
      count: records.length,
      total_employees: db.prepare('SELECT COUNT(*) as c FROM employees').get().c,
      date: today,
    });
  } catch (err) {
    res.status(500).json({ success: false, message: 'Failed to fetch today attendance' });
  }
});

// ─── GET /api/attendance/stats ────────────────────────────────────────────────
router.get('/stats', (req, res) => {
  try {
    const db = getDB();
    const today = new Date().toISOString().split('T')[0];
    const month = today.substring(0, 7);

    const totalEmployees = db.prepare('SELECT COUNT(*) as c FROM employees').get().c;
    const todayPresent   = db.prepare('SELECT COUNT(*) as c FROM attendance WHERE date = ?').get(today).c;
    const totalLogs      = db.prepare('SELECT COUNT(*) as c FROM attendance').get().c;
    const monthLogs      = db.prepare("SELECT COUNT(*) as c FROM attendance WHERE date LIKE ?").get(`${month}%`).c;

    res.json({
      success: true,
      data: {
        total_employees: totalEmployees,
        today_present: todayPresent,
        today_absent: totalEmployees - todayPresent,
        total_logs: totalLogs,
        this_month_logs: monthLogs,
      },
    });
  } catch (err) {
    res.status(500).json({ success: false, message: 'Failed to fetch stats' });
  }
});

module.exports = router;
